src/core/or/connection_or.c

```diff
@@ -765,8 +777,15 @@ connection_or_finished_connecting(or_connection_t *or_conn)
       return -1;
     }
 
-    connection_start_reading(conn);
     connection_or_change_state(or_conn, OR_CONN_STATE_PROXY_HANDSHAKING);
+
+    /* If the proxy type is haproxy, we don't have to wait for the response.
+     * We can start the TLS handshake immediately (but after all the outbound
+     * buffer is flushed). */
+    if (proxy_type != PROXY_HAPROXY) {
+      connection_start_reading(conn);
```

Comment:
confused here -- why should we not be reading in this case?

Reply:
Because after sending the PROXY header, we can do a tls handshake immediately and no need to wait for any response.

Reply:
Sure, but that doesn't mean we shouldn't listen for reads.  "start_reading()" just means that if information comes in, we'll notice and receive it, IIRC.